snapcraft/extensions/gnome.py

```diff
@@ -0,0 +1,264 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Generic GNOME extension to support core22 and onwards."""
+import dataclasses
+import functools
+import re
+from typing import Any, Dict, Optional, Tuple
+
+from overrides import overrides
+
+from .extension import (
+    Extension,
+    get_build_snaps,
+    get_extensions_data_dir,
+    prepend_to_env,
+)
+
+_SDK_SNAP = {"core22": "gnome-42-2204-sdk"}
+_PLATFORM_TRANSLATION = {"core22": "2204"}
+
+
+@dataclasses.dataclass
+class GNOMESnaps:
+    """A structure of GNOME related snaps."""
+
+    sdk: str
+    content: str
+    builtin: bool = True
+
+
+class GNOME(Extension):
+    """An extension that eases the creation of snaps that integrate with GNOME.
+
+    When used with core22 GNOME 42 will be used.
+
+    At build time it ensures the right build dependencies are setup and for
+    the runtime it ensures the application is run in an environment catered
+    for GNOME applications.
+
+    It configures each application with the following plugs:
+
+    - GTK3 Themes.
+    - Common Icon Themes.
+    - Common Sound Themes.
+    - The GNOME runtime libraries and utilities corresponding to 3.38.
+
+    For easier desktop integration, it also configures each application
+    entry with these additional plugs:
+
+    - desktop (https://snapcraft.io/docs/desktop-interface)
+    - desktop-legacy (https://snapcraft.io/docs/desktop-legacy-interface)
+    - gsettings (https://snapcraft.io/docs/gsettings-interface)
+    - opengl (https://snapcraft.io/docs/opengl-interface)
+    - wayland (https://snapcraft.io/docs/wayland-interface)
+    - x11 (https://snapcraft.io/docs/x11-interface)
+    """
+
+    @staticmethod
+    @overrides
+    def get_supported_bases() -> Tuple[str, ...]:
+        return ("core22",)
+
+    @staticmethod
+    @overrides
+    def get_supported_confinement() -> Tuple[str, ...]:
+        return "strict", "devmode"
+
+    @staticmethod
+    @overrides
+    def is_experimental(base: Optional[str]) -> bool:
+        return False
+
+    @overrides
+    def get_app_snippet(self) -> Dict[str, Any]:
+        return {
+            "command-chain": ["snap/command-chain/desktop-launch"],
+            "plugs": [
+                "desktop",
+                "desktop-legacy",
+                "gsettings",
+                "opengl",
+                "wayland",
+                "x11",
+            ],
+        }
+
+    @functools.cached_property
+    def gnome_snaps(self) -> GNOMESnaps:
+        """Return the GNOME related snaps to use to construct the environment."""
+        base = self.yaml_data["base"]
+        sdk_snap = _SDK_SNAP[base]
+        build_snaps = get_build_snaps(parts_yaml_data=self.yaml_data.get("parts"))
+        matcher = re.compile(r"gnome-\d+-" + _PLATFORM_TRANSLATION[base] + r"-sdk.*")
+        sdk_snap_candidates = [s for s in build_snaps if matcher.match(s)]
+        if sdk_snap_candidates:
```

Comment:
What happens if the user puts more than one -sdk snap? Is that checked elsewhere?